<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link rel="stylesheet" href="/stylesheets/styles.css">
    <link rel="stylesheet" href="/stylesheets/github-light.css">
    
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
<body>
<div>
<h2><a id="gill-lab-code-review" class="anchor" href="#gill-lab-code-review" aria-hidden="true"><span class="octicon octicon-link"></span></a><a href="/index.html">Gill lab Code Review</a></h2>

<h2>Topics:</h2>
<ol>
  <li>Homework assignment</li>
  <li>...</li>
</ol>

<h3>Homework 1: parsing .fastq files</h3>
<p>This week we will try a new format for our code review sessions. We will take a look at example files output from the Illumina MiSeq (courtesy of Andrew) and write small scripts to answer a few questions about the sequences. When we meet on Thursday, we will go over code snippets from the group.</p>

<ol>
  <li><a href="/downloads/miseq_data.zip">Download</a> the files. Open them in a text editor of your choice and briefly examine the format of each file.</li>
  <li>Build a dictionary in which the key corresponds to line 1 of the .fastq file (either file is fine) and the value corresponds to the raw sequence. We will talk more about how to build dictionaries and why they are useful on Thursday. For now, use this as an oportunity to read about and experiment with <a href="https://docs.python.org/2/tutorial/datastructures.html#dictionaries">python's dictionary implementation</a>.</li>
  <li>What is the average composition of each nucleotide (A, G, C, T) in the .fastq files? (Note: This information is not typically very useful unless you are looking for horizontal gene transfer events. The purpose of this exercise is to import and use the numpy.mean and numpy.std modules from the <a href="http://www.numpy.org/">NumPy</a> library. We will briefly talk about NumPy and <a href="http://matplotlib.org/">matplotlib</a> on Thursday.)</li>
  <li>Paired reads that overlap can be "merged" by aligning the forward and reverse reads to give a single .fastq or .fasta record for each pair. This merged file can then be split into separate files based on the barcodes and statistically analyzed to find something interesting about the experiment. Write a short script that uses <a href="http://drive5.com/usearch/">USEARCH</a> to build a merged .fastq file. Hint: the <a href="http://drive5.com/usearch/features_reads.html">USEARCH docs page</a> has some useful instructions.</li>
</ol>




    </div>
    <!--<script src="javascripts/scale.fix.js"></script>-->    
  </body>

<footer style="padding-top:50px;">
  <p><small>&copy;GillLab_CodeReview 2015</small></p>
</footer>

</html>